Add support for "make install"

rasky opened the pull request:
This patch adds support for installation of Tiny3D into `$N64_INST`. 

The main benefit of this is that it simplifies usage of Tiny3D in different projects as there is no more a `$T3D_INST` variable to set in the environment.

This PR creates a new file `t3d-inst.mk`, a variant that is installed as `$N64_INST/include/t3d.mk`. This file is extremely minimal: it just adds `-lt3d` to `LDFLAGS` and define a macro for the `$N64_INST/bin/gltf_to_t3d` tool. I'm honestly not even sure if this would be required anymore. We could simply part with the concept of a `t3d.mk` and let user Makefiles just add `N64_LDFLAGS += -lt3d` to link with Tiny3D.

For instance, this is the diff currently required fo a sample to build with the installed version of Tiny3D:

```diff
diff --git a/examples/12_uv_gen/Makefile b/examples/12_uv_gen/Makefile
index e60fd8d..f23f02f 100644
--- a/examples/12_uv_gen/Makefile
+++ b/examples/12_uv_gen/Makefile
@@ -2,7 +2,7 @@ BUILD_DIR=build
 T3D_INST=$(shell realpath ../..)

 include $(N64_INST)/include/n64.mk
-include $(T3D_INST)/t3d.mk
+include $(N64_INST)/include/t3d.mk

 N64_CFLAGS += -std=gnu2x -Os
```

Is the above much better than the following?

```patch
diff --git a/examples/12_uv_gen/Makefile b/examples/12_uv_gen/Makefile
index e60fd8d..bf2ad11 100644
--- a/examples/12_uv_gen/Makefile
+++ b/examples/12_uv_gen/Makefile
@@ -1,10 +1,9 @@
 BUILD_DIR=build
-T3D_INST=$(shell realpath ../..)

 include $(N64_INST)/include/n64.mk
-include $(T3D_INST)/t3d.mk

 N64_CFLAGS += -std=gnu2x -Os
+N64_LDFLAGS += -lt3d

 PROJECT_NAME=t3d_12_uv_gen

@@ -25,7 +24,7 @@ filesystem/%.sprite: assets/%.png
 filesystem/%.t3dm: assets/%.glb
        @mkdir -p $(dir $@)
        @echo "    [T3D-MODEL] $@"
-       $(T3D_GLTF_TO_3D) "$<" $@
+       $(N64_BINDIR)/gltf_to_t3d "$<" $@
        $(N64_BINDIR)/mkasset -c 2 -o filesystem $@

 $(BUILD_DIR)/$(PROJECT_NAME).dfs: $(assets_conv)
```

Obviously `t3d.mk` keeps the door open in case other changes are necessary. Not sure which would be, though.

thekovic commented:
I personally like the idea of abstracting the act of adding T3D to your Makefile behind a one-liner that includes `t3d.mk` so I'd be in favor of keeping it unless there are some obvious drawbacks to doing that

anacierdem commented:
Is this our new recommendation for any new "plugin" to libdragon? The convention is then having a `.mk` file in `$(N64_INST)/include`. There is one more thing we need to standardise and that is how to "install" the plugin. A root makefile makes sense but the tools one should also be included in that I believe.

Thinking on how this will be handled in the cli, it will create a submodule for the plugin, `make install` it and the code includes the `.mk` file to use the plugin library.

HailToDodongo commented:
Thanks for the changes!
I see 3 main scenarios to support:

1. Global install would be what this PR adds, so just an easy install globally, and users don't have to worry about other things like Env vars.

2. A local install: use-cases being pinning a version or having a custom modified t3d bound to a project, without having to globally re-install when switching projects.
This would almost work, but the headers now point to the wrong global directory without a .mk file.

3. Dev install: (working on t3d itself) is pretty much the same, just that the example projects are used to develop so they need to trigger a recompile of t3d.
This should be unaffected here.

I think i would want to keep the t3d.mk, and inside the file check if a T3D_INST is set or not.
If it is set, it would do the same things as now: setting the lib / include / gltf  path to T3D_INST.
If not set, it would only set the lib via N64_INST, and nothing else.

This should allow to drive which t3d install use via an env. from the outside, or makefile.
And it would be backwards compatible with users using t3d.mk + T3D_INST right now.

anacierdem commented:
One doubt I have is switching between local vs global library requires a change in the consuming makefile which will also be required to build a dockerized project even though t3d is apparently “local”. It would be best if we could have a single entrypoint. It is a little weird that docker users use the global method.

@HailToDodongo maybe you already mean to install a single `mk` that will be compatible with both based on the env. variable but won’t require a change going forward. Still the old installs with local mk will continue to work. Is thar what you suggest?

HailToDodongo commented:
Yes, there would be no braking change in both the PR as of now, and with a version that checks T3D_INST in t3d.mk.
In fact nothing would change at all by default for both new, old and docker users.
The fact that build.sh now always installs it globally is not an issue, since you are not forced to actually use it there

But it would allow you to now omit T3D_INST and letting it "fall-back" to the global install.

HailToDodongo commented:
Just noticed, this PR references `t3d-inst.mk` which doesn't exist
I misread as the current t3d.mk being installed under that name, if it doesn't then the T3D_INST check i mentioned would not be needed

rasky commented:
I've pushed an update. Now there are two supported installation:

**Global install**
Use `make install` to install Tiny3D into `N64_INST` (next to libdragon). Then a project can simply include the mk file to link Tiny3D:
```make
include $(N64_INST)/include/n64.mk
include $(N64_INST)/include/t3d.mk
```

**Local install**
If a project wants to have a local version of Tiny3D (for location modifications, or just pinning), it is sufficient to include the mk from the local copy:
```make
include $(N64_INST)/include/n64.mk
include path/to/t3d.mk
```
The local copy must still be built by itself, no dependency between the project and the library build system is added.

`T3D_INST` is no more. I've updated the README too.

I've renamed the library itself from `t3d.a` to `libt3d.a` as this allows to use the `-l` GCC option, which is more standard.

In the future it might be nice if the local installation mode also automatically rebuilt tiny3d if necessary, so that it is built just like it's part of the hosting project, but this can be a separate PR.

thekovic commented:
Only thing I would comment on here is how T3D headers are copied. Splitting them off into a separate folder and targetting that would be the more _proper_ way to do it but doing it like this is fine unless T3D starts growing a bunch more APIs in the future.

rasky commented:
Well, if you check your `/usr/include`, it is a mixed bag of headers for all dev packages you ever installed on your computer. The general idea of having a single prefix is that it is already part of your include path, so you can just include files from there, and even link library just with `-Lname` without having to have special paths (and/or .mk files) for each and every library you have installed in your computer.

thekovic commented:
Perhaps my previous comment wasn't clear enough. I was referring to splitting off headers on the side of Tiny3D's repo (making an `include` folder here), not on the installation folder side.

anacierdem commented on tools/gltf_importer/Makefile:
```diff
@@ -27,5 +27,8 @@ $(OBJDIR)/%.o: $(SRCDIR)/%.cpp
 gltf_to_t3d: $(OBJ)
 	$(CXX) $(CXXFLAGS) -o $@ $^ $ $(LINKFLAGS)
 
+install:
```
Should the install rules depend on the build as well?

rasky replied:
Yes, fixed
